lib/tests/streamlit/secrets_test.py

```diff
@@ -0,0 +1,129 @@
+db_username="Jane"
+db_password="12345qwerty"
+
+# Sub-sections are not loaded into os.environ
+[subsection]
+email="eng@streamlit.io"
+"""
+
+MOCK_SECRETS_FILE_LOC = "/mock/secrets.toml"
+
+
+class SecretsTest(unittest.TestCase):
+    def setUp(self) -> None:
+        # st.beta_secrets modifies os.environ, so we save it here and
+        # restore in tearDown.
+        self._prev_environ = dict(os.environ)
+        # Run tests on our own Secrets instance to reduce global state
+        # mutations.
+        self.secrets = Secrets(MOCK_SECRETS_FILE_LOC)
+
+    def tearDown(self) -> None:
+        os.environ.clear()
+        os.environ.update(self._prev_environ)
+
+    @patch_config_options({"server.fileWatcherType": "none"})
+    @patch("builtins.open", new_callable=mock_open, read_data=MOCK_TOML)
+    def test_access_secrets(self, _):
+        self.assertEqual(self.secrets["db_username"], "Jane")
+        self.assertEqual(self.secrets["subsection"]["email"], "eng@streamlit.io")
+
+    def test_secrets_file_location(self):
+        """Verify that we're looking for secrets.toml in the right place."""
+        self.assertEqual(os.path.abspath("./.streamlit/secrets.toml"), SECRETS_FILE_LOC)
+
+    @patch_config_options({"server.fileWatcherType": "none"})
+    @patch("builtins.open", new_callable=mock_open, read_data=MOCK_TOML)
+    def test_os_environ(self, _):
+        """os.environ gets patched when we load our secrets.toml"""
+        # We haven't loaded secrets yet
+        self.assertEqual(os.environ.get("db_username"), None)
+
+        # Load secrets
+        _ = self.secrets["db_username"]
+        self.assertEqual(os.environ["db_username"], "Jane")
+        self.assertEqual(os.environ["db_password"], "12345qwerty")
+
+        # Subsections do not get loaded into os.environ
+        self.assertEqual(os.environ.get("subsection"), None)
+
+    @patch_config_options({"server.fileWatcherType": "none"})
+    @patch("streamlit.error")
+    def test_missing_toml_error(self, mock_st_error):
+        """Secrets access raises an error, and calls st.error, if
+        secrets.toml is missing.
+        """
+        with patch("builtins.open", mock_open()) as mock_file:
+            mock_file.side_effect = FileNotFoundError()
+
+            with self.assertRaises(OSError):
+                self.secrets.get("no_such_secret", None)
+
+        mock_st_error.assert_called_once_with(
+            f"Secrets file not found. Expected at: {MOCK_SECRETS_FILE_LOC}"
+        )
+
+    @patch_config_options({"server.fileWatcherType": "none"})
+    @patch("builtins.open", new_callable=mock_open, read_data="invalid_toml")
+    @patch("streamlit.error")
+    def test_malformed_toml_error(self, mock_st_error, _):
+        """Secrets access raises an error, and calls st.error, if
+        secrets.toml is malformed.
+        """
+        with self.assertRaises(TomlDecodeError):
+            self.secrets.get("no_such_secret", None)
```

Comment:
nit: similarly would be slightly clearer calling `load_if_toml_exists` here

Reply:
In this case, `load_if_toml_exists` deliberately suppresses the st.error we're testing for at the end of the function